### PR TITLE
Use char32_t instead of uint32_t for UniChar

### PR DESCRIPTION
**Pull request by biodranik**
That allows to use std::u32string_view in the code, adding useful std::basic_string_view methods

### CODE REVIEW

File: base/base_tests/string_utils_test.cpp [hunk 1]
```diff
@@ -1270,3 +1271,18 @@ UNIT_TEST(Trim)
...
   TEST_EQUAL(str, "ri", ());
 }
+
+UNIT_TEST(UniChar_Ostream_Istream)
+{
+  strings::UniChar c1, c2, c3 = 0xabcdef;
+  std::istringstream istreamUpper("ABCDEF");
+  istreamUpper >> c1;
+  std::istringstream istreamLower("00abcdef");
+  istreamLower >> c2;
```
Comment (biodranik) [inline base/base_tests/string_utils_test.cpp:1281]: @vng we can read and convert unicode properly here, instead of reading from/to hex. WDYT?

**Comment by biodranik** [comment]
@zyphlar PTAL, this can make your code simpler.

File: indexer/indexer_tests/trie_test.cpp [hunk 2]
```diff
@@ -33,12 +33,12 @@ struct ChildNodeInfo
 
   uint32_t Size() const { return m_size; }
   bool IsLeaf() const { return m_isLeaf; }
-  uint32_t const * GetEdge() const { return &m_edge[0]; }
+  char32_t const * GetEdge() const { return &m_edge[0]; }
```
Comment (vng) [inline indexer/indexer_tests/trie_test.cpp:36]: use trie::TrieChar

File: indexer/indexer_tests/trie_test.cpp [hunk 3]
```diff
@@ -33,12 +33,12 @@ struct ChildNodeInfo
...
   size_t GetEdgeSize() const { return m_edge.size(); }
 
   bool m_isLeaf;
   uint32_t m_size;
-  vector<uint32_t> m_edge;
+  vector<char32_t> m_edge;
```
Comment (vng) [inline indexer/indexer_tests/trie_test.cpp:41]: use trie::TrieChar

File: indexer/trie.hpp [hunk 4]
```diff
@@ -11,12 +11,12 @@
...
 namespace trie
 {
-using TrieChar = uint32_t;
+using TrieChar = char32_t;
 
 // 95 is a good value for the default baseChar, since both small and capital latin letters
 // are less than +/- 32 from it and thus can fit into supershort edge.
 // However 0 is used because the first byte is actually language id.
-uint32_t constexpr kDefaultChar = 0;
+char32_t constexpr kDefaultChar = 0;
```
Comment (vng) [inline indexer/trie.hpp:19]: TrieChar

File: base/string_utils.hpp [hunk 5]
```diff
@@ -684,3 +684,9 @@ size_t EditDistance(Iter const & b1, Iter const & e1, Iter const & b2, Iter cons
   return prev[m];
 }
 }  // namespace strings
+
+// Reads hex-encoded character like 1FB2 or 01fb2 from a char stream.
+std::istream& operator>>(std::istream & inputStream, strings::UniChar & outChar);
+
+// Writes hex-encoded character like 1FB2 to a char stream.
+std::ostream& operator<<(std::ostream & outputStream, strings::UniChar const & outChar);
```
Comment (vng) [inline base/string_utils.hpp:692]: Where do we need it except tests?

Comment (biodranik) [reply base/string_utils.hpp]: Removed. PTAL

File: base/base_tests/string_utils_test.cpp [hunk 6]
```diff
@@ -46,9 +46,11 @@ UNIT_TEST(LowerUniChar)
     if (!semicolon)
       continue;
 
-    std::istringstream stream((std::string(*semicolon)));
-    strings::UniChar uc;
+    const std::string iString{*semicolon};
+    std::istringstream stream(iString);
```
Comment (vng) [inline base/base_tests/string_utils_test.cpp:50]: nit: What was wrong with previous ```std::istringstream stream((std::string(*semicolon)))``` ?

**Review by vng — APPROVED** [review]

**Comment by vng** [comment]
glyph_manager.cpp fix should be simple, just read uint32_t

File: drape/glyph_manager.cpp [hunk 7]
```diff
@@ -82,9 +82,8 @@ void ParseUniBlocks(std::string const & uniBlocksFile, ToDo toDo)
   while (true)
   {
     std::string name;
-    strings::UniChar start;
-    strings::UniChar end;
-    fin >> name >> std::hex >> start >> std::hex >> end;
+    uint32_t start, end;
+    fin >> std::hex >> name >> std::hex >> start >> end;
```
Comment (vng) [inline drape/glyph_manager.cpp:86]: You changed the sequence.

**Pull request merged**